iModelCore/ECDb/ECDb/ECDbValidationChecks.cpp

```diff
@@ -0,0 +1,143 @@
+    };
+
+    std::vector<CheckData> checks = {
+        {"nav_prop_id_check", ECDbValidationChecks::NavigationPropertyIdCheck},
+        {"class_id_check", ECDbValidationChecks::ClassIdCheck},
+    };
+
+    std::vector<ECDbValidationChecks::SummarizedResult> result;
+
+    bool continueChecks = true;
+
+    for(auto & check: checks) {
+        if(continueChecks) {
+            result.push_back ({
+                check.checkName,
+                check.checkFunction(ecdb)[0].status
+            });
+
+            if(result.back().status == "Failed") 
+                continueChecks = false;
+        }
+        else {
+            result.push_back ({
+                check.checkName,
+                "Did not run"
+            });
+        }
+    }
+
+    return result;
+}
+
+std::vector<ECDbValidationChecks::CheckResult> ECDbValidationChecks::NavigationPropertyIdCheck(ECDbCR ecdb) {
+    bool checkSuccessful = true;
+
+    std::vector<ECDbValidationChecks::CheckResult> result;
+
+    ECSqlStatement navigationValidityCheck;
+
+    navigationValidityCheck.Prepare(ecdb, "SELECT ec_classname(propdef.Class.id), propdef.Name FROM meta.ECPropertyDef as propdef LEFT JOIN meta.ECClassdef as classdef ON propdef.NavigationRelationshipClass.id = classdef.ECInstanceId WHERE propdef.NavigationRelationshipClass IS NOT NULL AND (classdef.ECInstanceId IS NULL OR classdef.Type != 1)");
+
+    struct PropertyData {
+        Utf8String className;
+        Utf8String name;
+    };
+
+    std::vector<PropertyData> properties = {};
+
+    while(BE_SQLITE_ROW == navigationValidityCheck.Step()) {
+        properties.push_back({navigationValidityCheck.GetValueText(0), navigationValidityCheck.GetValueText(1)});
+        }
+
+    if(properties.size() != 0) {
+        for(auto& property : properties) {
+            ecdb.GetImpl().Issues().ReportV(
+            IssueSeverity::Error,
+            IssueCategory::BusinessProperties,
+            IssueType::ECDbIssue,
+            "Could not find classId of navigation property %s in class %s", property.name.c_str(), property.className.c_str());
+
+            result.push_back({
+                "Failed",
+                Utf8PrintfString("Could not find classId of navigation property %s in class %s", property.name.c_str(), property.className.c_str()).c_str()
+                });
+    
+            checkSuccessful = false;
+            }
+        }
+
+    if(checkSuccessful)
+        {
+        result.push_back({
+            "Passed",
+            ""
+            });
+        return result;
+        }
+
+    return result;
+}
+
+std::vector<ECDbValidationChecks::CheckResult> ECDbValidationChecks::ClassIdCheck(ECDbCR ecdb) {
+    bool checkSuccessful = true;
+
+    std::vector<ECDbValidationChecks::CheckResult> result;
+
+    ECSqlStatement classDefStatement;
+
+    classDefStatement.Prepare(ecdb,
+    SqlPrintfString("SELECT DISTINCT ec_classname(ECClassDef.ECInstanceId, 's.c')" 
```

Comment:
You can just select form `ECClassDef`. I do not see why you need to select from `ClassHasAllBaseClasses` which will return redundant results and then apply `DISTINCT` to it.